### PR TITLE
[TSK-84] create unit tests for Services, Mappers and Controllers, also CI pipeline

### PR DESCRIPTION
**Pull request by Hoocki**

### CODE REVIEW

File: src/main/java/com/example/trello/model/dto/card/Card.java [hunk 1]
```diff
@@ -50,4 +51,17 @@ public void setDescription(final String description) {
...
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        final Card card = (Card) o;
+        return Objects.equals(id, card.id) && Objects.equals(name, card.name) && Objects.equals(description, card.description) && Objects.equals(createdAt, card.createdAt) && Objects.equals(updatedAt, card.updatedAt);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(id, name, description, createdAt, updatedAt);
+    }
```
Comment (adsukharev) [inline src/main/java/com/example/trello/model/dto/card/Card.java:65]: why did you add both these methods?

File: src/test/java/com/example/trello/service/impl/CardServiceImplTest.java [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+        //when
+        Card result = cardService.addCard(boardId, CARD_MODIFICATION);
+
+        //then
+        assertThat(result).isEqualTo(CARD_1);
+    }
+
+    @Test
+    void deleteCard() {
```
Comment (adsukharev) [inline src/test/java/com/example/trello/service/impl/CardServiceImplTest.java:148]: rename this method and check others to follow the pattern

File: pom.xml [hunk 3]
```diff
@@ -46,6 +46,11 @@
             <artifactId>spring-boot-starter-test</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>com.h2database</groupId>
+            <artifactId>h2</artifactId>
+            <scope>test</scope>
+        </dependency>
```
Comment (adsukharev) [inline pom.xml:53]: why do you need it?

Comment (Hoocki) [reply pom.xml:53]: i forgot delete this dependency

File: src/main/java/com/example/trello/model/dto/card/Card.java (hunk 1 above)
Comment (Hoocki) [reply src/main/java/com/example/trello/model/dto/card/Card.java:65]: in BoardMapperTest and CardMapperTest i want to compare instances using isEqualTo(), but if i didn't add these methods i have to compare their fields like: assertThat( result.getName() ).isEqualTo( expectedBoard.getName() )

File: src/test/java/com/example/trello/service/impl/CardServiceImplTest.java [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+        List<Card> expectedCards = List.of();
+        List<CardEntity> cardEntities = List.of();
+
+        given(cardRepository.findAllByBoardEntityIdOrderByCreatedAtDesc(boardId)).willReturn(cardEntities);
+
+        //when
+        List<Card> result = cardService.getCards(boardId);
+
+        //then
+        assertThat(result).isEqualTo(expectedCards);
```
Comment (adsukharev) [inline src/test/java/com/example/trello/service/impl/CardServiceImplTest.java:127]: after you update your import path, you will have more usefull methods. You can replace current assertion with `assertThat(result).isEmpty();`

File: src/test/java/com/example/trello/service/impl/CardServiceImplTest.java [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+
+        given(cardRepository.findAllByBoardEntityIdOrderByCreatedAtDesc(boardId)).willReturn(cardEntities);
+        given(cardMapper.map(CARD_ENTITY_1)).willReturn(CARD_1);
+        given(cardMapper.map(CARD_ENTITY_2)).willReturn(CARD_2);
+
+        //when
+        List<Card> result = cardService.getCards(boardId);
+
+        //then
+        assertThat(result).isEqualTo(expectedCards);
```
[the remark this answers is not in the document]
Comment (Hoocki) [reply src/test/java/com/example/trello/service/impl/CardServiceImplTest.java]: In services, I only check that the correct elements are returned, and in the repository tests I will check the order of the elements (since it is the repository that performs this function)

[the remark this answers is not in the document]
Comment (adsukharev) [reply src/test/java/com/example/trello/service/impl/CardServiceImplTest.java]: yes, but anyway you expect that this function returns you list in sorted order. Maybe later someone will make a change that break it by adding some other functionality

File: src/test/java/com/example/trello/controller/BoardControllerTest.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+
+    @InjectMocks
+    private BoardController boardController;
+
+    @Mock
+    private BoardService boardService;
+
+    private BoardEntity boardEntity1;
+
+    private BoardModification boardModification;
```
Comment (adsukharev) [inline src/test/java/com/example/trello/controller/BoardControllerTest.java:28]: what defenetely you don't need to do is to keep empty fields like this. Either remove it from here and use these vars inside of methods or create an object and use it everywhere if you want to make it common

File: github/workflows/ci.yaml
Comment (adsukharev) [inline github/workflows/ci.yaml:1]: You did not follow the instruction of how to create a pipeline.
How do you understand that your pipeline works and github runs your tests?

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v2
+
+      - name: Run tests
+        run: mvn test
+
+      - name: Print success message
+        run: echo "TEST SUCCESS"
```
Comment (adsukharev) [inline .github/workflows/ci.yml:28]: why do you need 2 almost the same steps?

File: src/test/java/com/example/trello/controller/BoardControllerTest.java [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+
+        given(boardService.getBoards()).willReturn(expectedBoards);
+
+        // when
+        List<BoardEntity> result = boardController.getBoards();
+
+        // then
+        assertThat(result)
+                .containsExactlyElementsOf(expectedBoards)
+                .containsExactlyInAnyOrder(boardEntity1, boardEntity2);
```
Comment (adsukharev) [inline src/test/java/com/example/trello/controller/BoardControllerTest.java:41]: why do you need 2 assertions ?

File: src/test/java/com/example/trello/mapper/BoardMapperTest.java [hunk 9]
```diff
@@ -0,0 +1,25 @@
+package com.example.trello.mapper;
+
+import com.example.trello.model.dto.board.BoardModification;
+import com.example.trello.model.entity.board.BoardEntity;
+import org.junit.jupiter.api.Test;
+import static org.assertj.core.api.Assertions.assertThat;
+
+class BoardMapperTest {
+
+    private BoardMapper BOARD_MAPPER = new BoardMapper();
```
Comment (adsukharev) [inline src/test/java/com/example/trello/mapper/BoardMapperTest.java:10]: add final to not get warnings

File: src/test/java/com/example/trello/service/impl/BoardServiceImplTest.java [hunk 10]
```diff
@@ -45,7 +46,7 @@ void should_returnBoards_whenBoardsExist() {
         //then
         assertThat(result)
                 .containsExactlyElementsOf(expectedBoards)
-                .containsExactlyInAnyOrder(boardEntity1, boardEntity2);
+                .isSortedAccordingTo(Comparator.comparing(BoardEntity::getName));
```
Comment (adsukharev) [inline src/test/java/com/example/trello/service/impl/BoardServiceImplTest.java:49]: why god, why?

Comment (adsukharev) [reply src/test/java/com/example/trello/service/impl/BoardServiceImplTest.java:49]: what does `containsExactlyElementsOf` does ?

File: src/test/java/com/example/trello/service/impl/CardServiceImplTest.java [hunk 11]
```diff
@@ -96,8 +97,10 @@ void should_returnUpdatedCard_whenCardExists() {
     void should_returnCards_whenCardsExist() {
         //given
         long boardId = 1L;
-        Card card1 = new Card(1L,"cardName1", "cardDesc1", LocalDateTime.now(), LocalDateTime.now());
-        Card card2 = new Card(2L, "cardName2", "cardDesc2", LocalDateTime.now(), LocalDateTime.now());
+        LocalDateTime cardDate1 = LocalDateTime.of(2020, 9, 20, 12, 30);
+        LocalDateTime cardDate2 = LocalDateTime.of(2020, 10, 20, 12, 30);
```
Comment (adsukharev) [inline src/test/java/com/example/trello/service/impl/CardServiceImplTest.java:101]: why do you need it

File: src/test/java/com/example/trello/controller/BoardControllerTest.java (hunk 8 above)
Comment (adsukharev) [reply src/test/java/com/example/trello/controller/BoardControllerTest.java:41]: ?

File: src/test/java/com/example/trello/service/impl/CardServiceImplTest.java (hunk 11 above)
Comment (Hoocki) [reply src/test/java/com/example/trello/service/impl/CardServiceImplTest.java:101]: to check that isSortedAccordingTo work correctly

File: src/test/java/com/example/trello/service/impl/BoardServiceImplTest.java (hunk 10 above)
Comment (adsukharev) [reply src/test/java/com/example/trello/service/impl/BoardServiceImplTest.java:49]: https://assertj.github.io/doc/#assertj-core-group-contains

File: src/test/java/com/example/trello/service/impl/CardServiceImplTest.java (hunk 11 above)
Comment (adsukharev) [reply src/test/java/com/example/trello/service/impl/CardServiceImplTest.java:101]: why do you need that the most useless method? if you have already `containsExactlyElementsOf` that does comparison automatically for you with order validation

File: src/test/java/com/example/trello/service/impl/BoardServiceImplTest.java (hunk 10 above)
Comment (Hoocki) [reply src/test/java/com/example/trello/service/impl/BoardServiceImplTest.java:49]: i thought that check only that list contains all elements in expectedList and not checking the order

File: src/test/java/com/example/trello/service/impl/CardServiceImplTest.java (hunk 11 above)
Comment (adsukharev) [reply src/test/java/com/example/trello/service/impl/CardServiceImplTest.java:101]: `containsExactlyElementsOf` or `containsExactly` do all what you need in your life. Test yourself by changing orders, amount of elements and so on

File: src/test/java/com/example/trello/service/impl/BoardServiceImplTest.java (hunk 10 above)
Comment (Hoocki) [reply src/test/java/com/example/trello/service/impl/BoardServiceImplTest.java:49]: i know that i can check if it method also checks the order, but starting from the name of the method, I thought it checks only the elements

Comment (adsukharev) [reply src/test/java/com/example/trello/service/impl/BoardServiceImplTest.java]: we discussed that you must test it yourself instead of making assumptions before pushing a commit

File: src/test/java/com/example/trello/service/impl/CardServiceImplTest.java (hunk 11 above)
Comment (Hoocki) [reply src/test/java/com/example/trello/service/impl/CardServiceImplTest.java:100]: but anyway to be clear with sorting, it would be better to manually enter the time

File: src/test/java/com/example/trello/controller/BoardControllerTest.java (hunk 8 above)
Comment (adsukharev) [reply src/test/java/com/example/trello/controller/BoardControllerTest.java:41]: ??

Comment (Hoocki) [reply src/test/java/com/example/trello/controller/BoardControllerTest.java:41]: i deleted the last assertion

File: src/test/java/com/example/trello/controller/CardControllerTest.java [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+
+        given(cardService.getCards(boardId)).willReturn(expectedCards);
+
+        // when
+        List<Card> result = cardController.getCardsByBoardId(boardId);
+
+        // then
+        assertThat(result)
+                .containsExactlyElementsOf(expectedCards)
+                .containsExactlyInAnyOrder(card1, card2);
```
Comment (adsukharev) [inline src/test/java/com/example/trello/controller/CardControllerTest.java:43]: it's not cool, that I need to spend my time on repeating the same thing

File: src/test/java/com/example/trello/service/impl/CardServiceImplTest.java (hunk 11 above)
Comment (adsukharev) [reply src/test/java/com/example/trello/service/impl/CardServiceImplTest.java:100]: I think it's useless because it's not job of the service. Repository is responsible for it. However, if you decide to use it you should be consistent and add dates to all cases where you expect sorting: controllers, services and so on

File: src/test/java/com/example/trello/controller/BoardControllerTest.java (hunk 8 above)
Comment (adsukharev) [reply src/test/java/com/example/trello/controller/BoardControllerTest.java:41]: in that case, why we both see it?

**Comment by adsukharev** [comment]
look at your PR and fix it correctly, tomorrow I will check it

**Review by adsukharev — APPROVED** [review]

**Review by kostya4201337 — APPROVED** [review]
